File: src/src_platform_swing/fmg/swing/draw/img/Animator.java
package fmg.swing.draw.img;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

import fmg.common.Pair;
import fmg.common.ui.ITimer;
import fmg.core.img.IAnimator;
import fmg.swing.utils.Timer;

public class Animator implements IAnimator, AutoCloseable {

   private static Animator _singleton;
   public static Animator getSingleton() {
      if (_singleton == null)
         _singleton = new Animator();
      return _singleton;
   }

   private Animator() {
      _subscribers = new HashMap<>();
      _timer = new Timer();
      _timer.setInterval(1000/60); // The number of frames per second
      _timer.setCallback(() -> {
         long currentTime = new Date().getTime();
         _subscribers.forEach((k,v) -> v.second.accept(currentTime - v.first));
      } );
   }

   private final ITimer _timer;
   private final Map<Object, Pair<Long /* start time of subscribe */, Consumer<Long /* time from the beginning of the subscription */>>> _subscribers;

   @Override
   public void subscribe(Object subscriber, Consumer<Long /* time from start subscribe */> subscriberCallbackMethod) {
      _subscribers.put(subscriber, new Pair<>(new Date().getTime(), subscriberCallbackMethod));
   }

   @Override
   public void unsubscribe(Object subscriber) {
      _subscribers.remove(subscriber);
   }

   @Override
   public void close() {
      _timer.setCallback(null);
      _timer.close();
      _subscribers.clear();
   }

}

File: src/src_platform_android/fmg/android/img/Animator.kt
package fmg.android.img

import java.util.Date
import java.util.HashMap
import java.util.function.Consumer

import fmg.common.ui.ITimer
import fmg.core.img.IAnimator
import fmg.android.utils.Timer

class Animator private constructor() : IAnimator, AutoCloseable {

    private class SubscribeInfo {
        var active = true    // enabled?
        var startTime = Date().time // start time of subscribe
        lateinit var callback: Consumer<Long/* time from the beginning of the subscription */>
    }

    private val _timer: ITimer
    private val _subscribers: MutableMap<Any, SubscribeInfo/* subscriber */>

    init {
        _subscribers = HashMap()
        _timer = Timer()
        _timer.interval = (1000 / 60).toLong() // The number of frames per second
        _timer.setCallback {
            val currentTime = Date().time
            _subscribers.forEach { k, v ->
                if (v.active)
                    v.callback.accept(currentTime - v.startTime)
            }
        }
    }

    override fun subscribe(subscriber: Any, subscriberCallbackMethod: Consumer<Long/* time from start subscribe */>) {
        var info: SubscribeInfo? = _subscribers[subscriber]
        if (info == null) {
            info = SubscribeInfo()
            info.callback = subscriberCallbackMethod
            _subscribers[subscriber] = info
        } else {
            info.active = true
            info.startTime = Date().time - info.startTime // apply of pause delta time
        }
    }

    override fun pause(subscriber: Any) {
        val info = _subscribers[subscriber] ?: return
        info.active = false
        info.startTime = Date().time - info.startTime // set of pause delta time
    }

    override fun unsubscribe(subscriber: Any) {
        _subscribers.remove(subscriber)
    }

    override fun close() {
        _timer.setCallback(null)
        _timer.close()
        _subscribers.clear()
    }

    companion object {

        private var _singleton: Animator? = null // not synchronized. since should work only in the thread of the UI.

        val singleton: Animator
            get() {
                if (_singleton == null)
                    _singleton = Animator()
                return _singleton!!
            }
    }

}
